src/hooks/jukebox.ts

```diff
@@ -0,0 +1,127 @@
+import { isEqual } from "lodash";
+
+import { VENUE_CHAT_AGE_DAYS } from "settings";
+
+import { sendJukeboxMessage } from "api/jukebox";
+
+import { JukeboxMessage, SendJukeboxMessage } from "types/jukebox";
+
+import {
+  getBaseMessageToDisplay,
+  partitionMessagesFromReplies,
+} from "utils/chat";
+import { WithId } from "utils/id";
+import { buildJukeboxMessage } from "utils/jukebox";
+import { jukeboxMessagesSelector } from "utils/selectors";
+import { getDaysAgoInSeconds } from "utils/time";
+import { isTruthy } from "utils/types";
+
+import { useFirestoreConnect } from "hooks/useFirestoreConnect";
+import { useWorldUsersByIdWorkaround } from "hooks/users";
+import { useSelector } from "hooks/useSelector";
+import { useUser } from "hooks/useUser";
+
+const noMessages: WithId<JukeboxMessage>[] = [];
+
+export const useJukeboxChat = ({
+  venueId,
+  tableId,
+}: {
+  venueId?: string;
+  tableId?: string;
+}) => {
+  const messagesToDisplay = useJukeboxMessages(venueId, tableId);
+
+  const { sendJukeboxMsg } = useJukeboxActions(venueId, tableId);
+
+  return {
+    messagesToDisplay,
+    sendJukeboxMsg,
+  };
+};
+
+const useJukeboxActions = (venueId?: string, tableId?: string) => {
+  const { userId } = useUser();
+
+  const sendJukeboxMsg: SendJukeboxMessage = useCallback(
+    async ({ message, url }) => {
+      if (!venueId || !userId || !tableId) return;
+      const processedMessage = buildJukeboxMessage<JukeboxMessage>({
+        from: userId,
+        text: message,
+        url: url,
+        tableId,
+      });
+      return sendJukeboxMessage({
+        venueId,
+        tableId,
+        message: processedMessage,
+      });
+    },
+    [venueId, userId, tableId]
+  );
+
+  return {
+    sendJukeboxMsg,
+  };
+};
+
+const useJukeboxMessages = (venueId?: string, tableId?: string) => {
+  const { worldUsersById } = useWorldUsersByIdWorkaround();
+  const { userId } = useUser();
+
+  useConnectVenueJukeboxMessages(venueId, tableId);
+
+  const chatMessages =
+    useSelector(jukeboxMessagesSelector, isEqual) ?? noMessages;
+
+  const venueChatAgeThresholdSec = getDaysAgoInSeconds(VENUE_CHAT_AGE_DAYS);
+
+  const filteredMessages = useMemo(
+    () =>
+      chatMessages.filter(
+        (message) => message.ts_utc.seconds > venueChatAgeThresholdSec
+      ),
+    [chatMessages, venueChatAgeThresholdSec]
+  );
+
+  const { messages } = useMemo(
+    () => partitionMessagesFromReplies(filteredMessages),
+    [filteredMessages]
+  );
+
+  return useMemo(
+    () =>
+      messages
+        .map((message) => {
+          const displayMessage = getBaseMessageToDisplay<
+            WithId<JukeboxMessage>
+          >({
+            message,
+            usersById: worldUsersById,
+            myUserId: userId,
+          });
+
+          if (!displayMessage) return undefined;
+
+          return { ...displayMessage };
+        })
+        .filter(isTruthy),
+    [userId, worldUsersById, messages]
+  );
+};
```

Comment:
This logic looks very similar to the one in `venueChat.ts`. Messages there are also first filtered by query, then filtered by age and then partitioned by replies and transformed into `messagesToDisplay` form. Would be nice if this was deduplicated as we may change the way we deal with chat messages in the future and if this would be deduplicated it would certainly help!

Reply:
yeah, I agree that the logic is the same there, altough some pieces are missing for jukebox messages as there's not that much business logic in there
i'd say that with some of the filtering removed in d0f3e28 there's not much to deduplicate
even though they share some logic the entities are different and should be treated so